models/Task.php

```diff
@@ -1143,4 +1148,16 @@ private function getWorker()
     {
         return User::findOne(['id' => $this->getAccount(Task::ACCOUNT_WORKER)->user_id])->guid;
     }
+
+    /**
+     * Delete all transactions where Task Space Account is the target account
+     *
+     * @param $account
+     */
+    private function refund($account){
+        $incomeTransactions = Transaction::findAll(['to_account_id' => $account->id]);
+        foreach ($incomeTransactions as $transaction){
+            $transaction->delete();
```

Comment:
Why deleting transactions, couldn't you just initiate another refund transaction? I think transactions shouldn't be deleted ever.. or if the logic is indicating that you must delete a transaction, then the function shouldn't be called refund :wink:

Reply:
I was reverting the transactions for the first time that's why you are seeing the word "refund", but i've faced a problem when deleting the task, in this case will have *null foreign keys* of the deleted task account in transactions records. 
Keep in mind that humhub is not using softdeleteable feature ( a lot of work to do there ), i'm making it simple for a first version. 

When a task is not completed and the worker target account is not involved yet, i think it's not a big deal when deleting a task to revert transactions by a cascade delete. While Once the task is completed and the payment for the target account is done, no way to delete transactions that's why for each made transaction i'm changing *the task account id* foreign key by *space default account id* to keep balances the way they are (see mutateTransactions method). I'm not using this in the refund method because it's not logic to have the same account as sender and recipient at the same time.

Reply:
@KarimChabrak please follow this, because this is all the things where those "simple" things suddenly get complicated when thought through and used. 

would removing the possibility to delete a task help?

Reply:
@KarimChabrak what do you think ?

Reply:
@gdalyy can't you just disable them instead of deleting? either showing it disabled or not (fake delete) for users?

Reply:
As i mentionned using softdeleteable extension may take alot of time, i did'nt work with such feature with yii2 before. I'll work around anyway.

Reply:
let's see what karim will say first

Reply:
When I try to delete a task, I see that the coins are transferred back to initial account and all accounts and transactions related to the task are deleted.

From accounting perspective, I do not see the problem. It is a general problem how can people allocating task and people who started working on a task track their agreements and dicsussions. 

Deleting is good if all people agree as old not relevant stuff is removed, but if there are conflicts we need to track what happened. 

My proposal is to leave this as it is now. If there are conflicts, we should open a new issue which prevents deletion and optionally introduces a kind of archive.

Reply:
To be clear @KarimChabrak , i'm not deleting transactions everytime a task is deleted , i'm just deleting them when a task which is not completed or not started is deleted. 
Once a task is completed then it's deleted , i'm not removing any transaction, i'm just changing the task account ID present in the transaction record by the space default accout ID to keep balances the way they are.

Reply:
I did not fully understand, but it seems to work.
I would tend to not allow deletion or change of task content as soon coins are transferred to the assigned user account go ensure that a communix can track why coins did move out. If this is an additional work, leave it for the moment untill we open a new issue. 

I think we can deploy the feature now in our life network.

Reply:
@craxrev can you approve changes and merge the PR ?

Reply:
@KarimChabrak  I guess the main point for you to see here is, that there are uncounable edge cases that have implication on how the system works even though you might not see it when you finish or delete one single task.